### ISSUE TITLE
Iteration Plan for Nov 2019

### ISSUE THREAD

**Issue by rabbit008** [microsoft/nni#1549]
**Feature freeze date**: ~11/8~ 11/15
**Code freeze date**: ~11/15~ 11/22
**Target Code Freeze Date**: ~11/25~ 12/02
**Release Manager**: @chicm-ms 

### 1. PAI Integration (@SparkSnail @yds05 ) - 
- [x] (1) Token integration PR Review Done, ready to merge @SparkSnail ; 
        - [ ] **Test owner**: @chicm-ms   account issue

### 2. GPU related issue (@liuzhe-lz @chicm-ms @QuanluZhang ) - 
- [x]  go through GPU-related issues (79 -> 2) @ultmaster : https://github.com/microsoft/nni/search?q=GPU&type=Issues
- [x]  permission issue of gpu_metric_collector.sh; @liuzhe-lz  1 day  ~11/4~ 11/8 (hotfix PR ready for review, reviewers: @SparkSnail @QuanluZhang ) to fix multiple user issue for remote machine
        - [X] **Test owner**: @QuanluZhang 
- [x] scheduling issue in remote mode (not evenly distributed among machines); @chicm-ms  1 day next week  11/5 (PR ready for review, reviewers: @SparkSnail @liuzhe-lz )
        - [X] **Test owner**: @tanglang96 
- [x]  [need design] tell user no available gpus (give more hint); @quanlu 1 day  to be updated ~10/30~ ,  11/1   design review 11/1 (**DONE**: design ready for review @QuanluZhang )

### 3. Test (@ultmaster @chicm-ms )
- [x] more test for tuners; 
- [x] Windows  PR pipeline  11/1
- [x]  more test for installation @chimin 3 day,  to be started ~10/29~,  11/1 (PR ready, reviewers: @liuzhe-lz @SparkSnail )
- [x] fpgm pruner unit test 11/18, PR

### 4. NAS (@ultmaster @QuanluZhang  @liuzhe-lz)
- [X]  API finalize review – ~10.23~ @Quanlu Zhang  review meeting 10/29
- [X]  Design new APIs in nni sdk for NAS  TBD 1 week,  11/1 to finalized design
- [x]  Support darts, enas, following the new interface design ,  @yuge ~11/8~, 11/18 to merge.
- [x]  Model export @yuge  PR 11/15
- [X]  ~one-shot NAS~,  pdarts @chi  ~11/8~ **ETA 11/15 PR,  PR #1730 submitted 11/13 @squirrelsc 

### 5. Model Compression (@tanglang96 @chicm-ms )
- [x]  Add tutorial for auto model compression  @tanglang96    1 day (Done)
        - [ ] **Test owner**: @scarlett2018 
- [x]  Add 2 new pruning algorithms (slim pruner, filter pruner) Lang 
        - [X] **Test owner**: @liuzhe-lz 
- [X]  FPGM pruner, @chicm-ms  
        - [X] **Test owner**: @tanglang96 
- [x]  Lottery ticket pruner  @QuanluZhang   PR submitted 11/1, need refactor. **ETA 11/8**, to support tf2.0
        - [X] **Test owner**: @chicm-ms 
- [x]  Compressor framework refactor @QuanluZhang  11/1
- [X]  TF 2.0 compressor implementation @chicm-ms   ~11/8~ **ETA 11/15 @chicm-ms** 
- [x]  Add API for exporting compressed model 3 day @tanglang96  ,  ~11/1~  11/6,  PR submitted, reviewers:  @liuzhe-lz @QuanluZhang @chicm-ms 
        - [X] **Test owner**: @chicm-ms 
- [X]  Add ~2~ 1 quantization algo @Shanbin 11/8,  implemented one algo(xupu,  to double check),  first algo implemented, to submit PR 11/4,  finished cifar10 test,  to test on imagenet
        - [X] **Test owner**: @QuanluZhang 
- [X] Quantization refactor @Shanbin

### 7. Customized trial 
- [X] @liuzhe-lz – [code ready for next release], UI PR submited,  tag PR ready ~10/30~, ~11/1 PR~, ETA 11/8    webui @lijiao  PR #1653,  reviewers: @liuzhe-lz  @chicm-ms 
       - [ ] **Test owner**: @ultmaster @lvybriage 
- [X] blocking issue - trial auto killed after added more than 3 same customized trial. @liuzhe-lz 

### 8. Web UI changes (@lvybriage ) – pick up items from web ui backlog
  - [X] **Test owner**: @QuanluZhang 

- [X] chart zoom out/in,  week of ~11/4~,  ~11/8~,  PR 11/13
- [X]  Fix bug:  min/max settings refresh issue @lvybriage  11/4

### 9. Tensorflow 2.0 support (@liuzhe-lz )
- [X]  Design for TensorFlow 2.0 supports @liuzhe-lz 11/11 ~ 11/13
- [x]  Mnist tf2.0 tf.keras example 11/18 @liuzhe-lz 
   - [x] **Test owner**: @SparkSnail (hold till IT pass)

### ~10. Rocks db example system auto tuning  (@gxiaotian,  @Xuehui )  [Done]~
  - [ ] **Test owner**: @QuanluZhang talk with @gxiaotian to test

### 11. bugs - @ultmaster  
- [x] go through current bug list and schedule a review with team for triage  (Go through Done)

### ~12. pylint   : to use master config~
- [x] pylint pipeline  @chicm-ms 
- [x]  ppo tuner, smac tuner, @QuanluZhang,  PR submitted 
####  nni tools:
- [x]  annotation: @liuzhe-lz    PR submitted 
- [x]  nni_cmd:  @SparkSnail    10/30 to PR
- [x]  nni_gpu:  @Shanbin,  
- [x] nni_trial :  @tanglang96   PR submitted
- [x] pipeline numpy ignore @liuzhe-lz 11/4 
- [X]  pipeline dependencies @chicm-ms 
- [x] Fix windows pipeline  @chicm-ms 11/4
- [x] nnicli:  @chicm-ms  11/1

### 13. docstring: TBD
- [X] Trial   @chicm-ms      ~11/2~ to PR,   ~11/5~ 11/8,  PR submitted
#### Tuner  , Assessor
- [x] base tuner:  @liuzhe-lz   pr submitted
- [x] batch tuner,  evolution, hpo, networkmorphism, metis:  @xuehui1991   PR submitted, (pylint included)  pr submitted  reviewers:  @liuzhe-lz   @leckie-chn 
- [x] grid, ppo, smac,  : @QuanluZhang  11/1   waiting for Zhe's docstring PR 
- [x] gp tuner, medianstop assessor, curvfitting assessor: @suiguoxin   11/4, ready for merge
#### Advisor:   to be refined, next release

- [x] Liuzhe's pylint PR:   @leckie-chn review

### Deferred to next release
-----------------------------------------------
- [ ] (2) NFS support (PAI - Dian Wang) – **ETA ~11/15~ PR @SparkSnail**  pending on PAI mount implementation  
- [ ]  ~[only design] add new APIs in NNI SDK for supporting weight/checkpoint sharing (@liuzhe-lz ) , together with last item~,

### 6. Log and storage (@lvybriage  @liuzhe-lz )
- [ ]  view log on web ui (@lijiao) 3 day   (~11/1~, 11/15 to submit PR )  downloading tool and rest service support (@liuzhe-lz ) 3 day    ~week of 11/4~   draft PR submitted, reviewers: chicm, sparksnail

- [ ]  Proxyless NAS @QuanluZhang   ~11/8~ **ETA ~11/15~ PR @QuanluZhang**  11/18 PR

### Other candidates (not discussed)
14. installation code
15. [need detail design] Advisor Refactor
16. new tuning algo
17. tensorboard support
18. new training service
